File: 22.백트래킹/15649.N과 M (1)/main.py
import sys

input = sys.stdin.readline


def backtrack(n, lst):
    if n == M:
        print(*lst)
        return

    for i in range(1, N + 1):
        if visited[i] == True:
            continue

        visited[i] = True
        backtrack(n + 1, lst + [i])
        visited[i] = False


if __name__ == "__main__":
    N, M = map(int, input().split())
    visited = [False] * (N + 1)

    backtrack(0, [])

File: 22.백트래킹/15649.N과 M (1)/main.ts
const rl = require("readline").createInterface({
  input: process.stdin,
  output: process.stdout,
});

const inputs: number[] = [];

rl.on("line", (line: string) => {
  const values = line.split(" ").map(Number);
  inputs.push(...values);
}).on("close", () => {
  const N = inputs[0];
  const M = inputs[1];
  generatePermutations(N, M);
  process.exit();
});

function generatePermutations(N: number, M: number) {
  const visited: boolean[] = new Array(N + 1).fill(false);
  const permutation: number[] = [];

  function backtrack() {
    if (permutation.length === M) {
      console.log(permutation.join(" "));
      return;
    }

    for (let i = 1; i <= N; i++) {
      if (!visited[i]) {
        visited[i] = true;
        permutation.push(i);
        backtrack();
        visited[i] = false;
        permutation.pop();
      }
    }
  }

  backtrack();
}
